src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,431 @@
+	}
+
+	distributedTransctionId = GetCurrentDistributedTransctionId();
+
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	values[0] = ObjectIdGetDatum(MyDatabaseId);
+	values[1] = UInt64GetDatum(distributedTransctionId->initiatorNodeIdentifier);
+	values[2] = UInt64GetDatum(distributedTransctionId->transactionId);
+	values[3] = TimestampTzGetDatum(distributedTransctionId->timestamp);
+
+	tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * GetAllActiveDistributedTransactions returns all the active distributed
+ * transactions on the shared memory.
+ *
+ * Note that the function does not gurantee a global consistent result, the function
+ * only locks each active backend while copying it. Currently the only caller of this
+ * function is in the src/test folder. The consistency gurantee is sufficient for
+ * testing.
+ */
+List *
+GetAllActiveDistributedTransactions(void)
+{
+	int sessionIndex = 0;
+	List *activeTransactionList = NIL;
+
+	for (sessionIndex = 0; sessionIndex < MaxBackends; ++sessionIndex)
+	{
+		DistributedTransactionBackendData *currentBackend =
+			&distributedTransactionShmemData->sessions[sessionIndex];
+
+		if (currentBackend && currentBackend->transactionId.transactionId != 0)
+		{
+			DistributedTransactionBackendData *activeBackend =
+				(DistributedTransactionBackendData *)
+				palloc(sizeof(DistributedTransactionBackendData));
+
+			SpinLockAcquire(&currentBackend->mutex);
+
+			activeBackend->databaseId = currentBackend->databaseId;
+			activeBackend->transactionId.initiatorNodeIdentifier =
+				currentBackend->transactionId.initiatorNodeIdentifier;
+			activeBackend->transactionId.transactionId =
+				currentBackend->transactionId.transactionId;
+			activeBackend->transactionId.timestamp =
+				currentBackend->transactionId.timestamp;
+
+			SpinLockRelease(&currentBackend->mutex);
+
+			activeTransactionList = lappend(activeTransactionList, activeBackend);
+		}
+	}
+
+	return activeTransactionList;
+}
+
+
+/*
+ * InitializeDistributedTransactionManagement requests the necessary shared memory
+ * from Postgres and sets up the shared memory startup hook.
+ */
+void
+InitializeDistributedTransactionManagement(void)
+{
+	/* allocate shared memory */
+	RequestAddinShmemSpace(DistributedTransactionManagementShmemSize());
+
+	prev_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = DistributedTransactionManagementShmemInit;
+}
+
+
+/*
+ * DistributedTransactionManagementShmemInit is the callback that is to be called on
+ * shared memory startup hook. The function sets up the necessary shared memory segment
+ * for the distributed transaction manager.
+ */
+static void
+DistributedTransactionManagementShmemInit(void)
+{
+	bool alreadyInitialized = false;
+	char *trancheName = "Distributed Transaction Management";
```

Comment:
String constants are constant. You shouldn't assign them to a non-const qualified pointer (it's disallowed in later versions of C++ too)